delivery_ups_oca/models/ups_request.py

```diff
@@ -0,0 +1,287 @@
+            "Dimensions": {
+                "UnitOfMeasurement": {"Code": self.package_dimension_code},
+                "Length": str(self.default_packaging_id.packaging_length),
+                "Width": str(self.default_packaging_id.width),
+                "Height": str(self.default_packaging_id.height),
+            },
+            "PackageWeight": {
+                "UnitOfMeasurement": {"Code": self.package_weight_code},
+                "Weight": str(PackageWeight),
+            },
+        }
+
+    def _prepare_rate_shipment(self, order):
+        packages = [self._quant_package_data_from_order(order)]
+        return {
+            "RateRequest": {
+                "Shipment": {
+                    "Shipper": self._partner_to_shipping_data(
+                        partner=order.company_id.partner_id,
+                        ShipperNumber=self.shipper_number,
+                    ),
+                    "ShipTo": self._partner_to_shipping_data(order.partner_shipping_id),
+                    "ShipFrom": self._partner_to_shipping_data(
+                        order.warehouse_id.partner_id or order.company_id.partner_id
+                    ),
+                    "Service": {"Code": self.service_code},
+                    "Package": packages,
+                }
+            }
+        }
+
+    def _rate_shipment(self, order, skip_errors=False):
+        status = self._process_reply(
+            url="%s/ship/v1807/rating/Rate" % self.url,
+            data=self._prepare_rate_shipment(order),
+        )
+        self._raise_for_status(status, skip_errors)
+        return status
+
+    def test_call(self, order):
+        res = self._rate_shipment(order, True)
+        return res["response"] if "response" in res else res
+
+    def rate_shipment(self, order):
+        status = self._rate_shipment(order)
+        return status["RateResponse"]["RatedShipment"]["TotalCharges"]
+
+    def _prepare_shipping_label(self, carrier_tracking_ref):
+        return {
+            "LabelRecoveryRequest": {
+                "LabelSpecification": self._label_data(),
+                "TrackingNumber": carrier_tracking_ref,
+            }
+        }
+
+    def shipping_label(self, carrier_tracking_ref):
+        status = self._process_reply(
+            url="%s/ship/v1807/shipments/labels" % self.url,
+            data=self._prepare_shipping_label(carrier_tracking_ref),
+        )
+        self._raise_for_status(status, False)
+        return status
+
+    def cancel_shipment(self, pickings):
+        url = "%s/ship/v1/shipments/cancel" % self.url
+        for item in pickings.filtered(lambda x: x.carrier_tracking_ref):
+            url = "{}/{}".format(url, item.carrier_tracking_ref)
+            status = self._process_reply(url=url, method="delete")
+            self._raise_for_status(status, False)
+        return True
+
+    def tracking_state_update(self, picking):
```

Comment:
This seems not to be working with production mode (it is working in test mode though). For some reason, the API sends an error. We have got in touch with UPS to try to sort it out, but we are still waiting for a solution. Are you dealing with this issue too? In that case, we can try to work together on this so it can be solved quickly. Thank you.

Reply:
Regarding UPS documentation https://developer.ups.com/api/reference?loc=en_US#operation/getSingleTrackResponseUsingGET it seems that Oauth is required now. That's why maybe it is working on sandbox environment but not in production.

@manuelregidor 

https://developer.ups.com/api/reference/oauth/authorization-code?loc=en_US

Reply:
OK, thank you.